### PR TITLE
Enable test_distributed to work with spawn mode

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42932 Rename test_distributed to test_distributed_fork
* **#41769 Enable test_distributed to work with spawn mode**

Closes https://github.com/pytorch/pytorch/issues/36663. Currently the tests in `test_distributed` only work with the `fork` mode multiprocessing, this PR introduces support for `spawn` mode multiprocessing as well (while keeping the `fork` mode intact).

Motivations for the change:
1) Spawn multiprocessing is the default on MacOS, so it better emulates how MacOS users would use distributed
2) With python 3.8+, spawn is the default on linux, so we should have test coverage for this
3) PT multiprocessing suggests using spawn/forkserver over fork, for sharing cuda tensors: https://pytorch.org/docs/stable/multiprocessing.html
4) Spawn is better supported with respect to certain sanitizers such as TSAN, so adding this sanitizer coverage may help us uncover issues.

How it is done:
1) Move `test_distributed` tests in `_DistTestBase` class to a shared file `distributed_test` (similar to how the RPC tests are structured)
2) For `Barrier`, refactor the setup of temp directories, as the current version did not work with spawn, each process would get a different randomly generated directory and thus would write to different barriers. 
3) Add all the relevant builds to run internally and in OSS.
Running test_distributed with spawn mode in OSS can be done with:
`python test/run_test.py -i distributed/test_distributed_spawn -v`

Differential Revision: [D22408023](https://our.internmc.facebook.com/intern/diff/D22408023/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22408023/)!

### CODE REVIEW

File: test/distributed/test_distributed_spawn.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            os.environ["MASTER_PORT"] = str(MASTER_PORT)
+            os.environ["WORLD_SIZE"] = str(WORLD_SIZE)
+            super().setUpClass()
+
+        def setUp(self):
+            super().setUp()
+            # initialize Barrier.
+            initialize_temp_directories()
+            Barrier.init()
+            self._spawn_processes()
```
Comment (pritamdamania87) [inline test/distributed/test_distributed_spawn.py:61]: Don't we only need to override the `setUp` method in this class? Shouldn't the other methods move to the base class and be shared between fork and spawn?

File: test/run_test.py [hunk 2]
```diff
@@ -91,6 +92,8 @@
     'distributed/rpc/test_process_group_agent',
     'distributed/rpc/test_tensorpipe_agent',
     'distributed/test_distributed',
+    'distributed/test_distributed_spawn',
+    'distributed/test_ddp_under_dist_autograd',
```
Comment (pritamdamania87) [inline test/run_test.py:96]: Why ddp_under_dist_autograd?

**Review by pritamdamania87 — COMMENTED** [review]
Awesome work, thanks for enabling these tests!

File: test/run_test.py (hunk 2 above)
Comment (rohan-varma) [reply test/run_test.py:96]: Looks like I incorrectly resolved some merge conflicts (here and below this file), this was deleted as part of the RPC test refactor in https://github.com/pytorch/pytorch/pull/40818. Will update to correct this.

File: test/distributed/test_distributed_spawn.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/test_distributed_spawn.py:61]: Good point, it looks like we can dedupe most of this with the fork version and put the shared code in `_DistTestBase`.

**Comment by rohan-varma** [comment]
@pritamdamania87 Would be great to get another look at this whenever you get a chance!

**Review by pritamdamania87 — APPROVED** [review]

**Pull request closed without merging**